Node-based @deprecated checks

sandersn opened the pull request:
Switch the checker to syntactic checks for `@deprecated` on declarations. This requires a bit more checking of declarations in the
checker at times, but it:

1. Gets rid of work, and a symbol flag, in the binder.
2. Skips work in the checker unless there is a `@deprecated` tag.
3. Makes it fairly simple to only issue errors on particular signatures of overloaded functions.

Fixes #39212, alternative to #39221 -- and tests are based on #39221's.

Edit: This now fixes part of #39374; JSX elements, decorators and tagged template functions are all calls and were not being marked deprecated.

sandersn commented:
@typescript-bot perf test this

sandersn commented:
Note: `isTypeDeclaration` from #39122 is better and should be copied here. I think the tests from here have a few more cases than those in #39122 and should be copied there.

Kingwl commented:
@typescript-bot pack this.

Kingwl commented on src/compiler/checker.ts:
```diff
@@ -13274,13 +13274,19 @@ namespace ts {
         }
 
+        function isUncalledFunctionReference(node: Node, symbol: Symbol) {
+            return !(symbol.flags & SymbolFlags.Function)
```
Maybe Method or Constructor?

Kingwl commented on src/compiler/checker.ts:
```diff
@@ -13274,13 +13274,19 @@ namespace ts {
 
+        function isUncalledFunctionReference(node: Node, symbol: Symbol) {
+            return !(symbol.flags & SymbolFlags.Function)
+                || !isCallExpression(findAncestor(node, n => !isAccessExpression(n)) ?? node.parent)
```
NewExpression?

DanielRosenwasser replied:
Or an invocationexpression in general

Just switch to `||`, it's not a primitive type

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25013,7 +25018,7 @@ namespace ts {
             }
             else {
-                if (prop.flags & SymbolFlags.Deprecated) {
+                if (prop.valueDeclaration?.flags & NodeFlags.Deprecated && isUncalledFunctionReference(node, prop)) {
```
This is potentially `undefined & NodeFlags.Deprecated` - is that intentional?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13274,13 +13274,19 @@ namespace ts {
                 const prop = getPropertyOfType(objectType, propName);
                 if (prop) {
-                    if (accessNode && prop.flags & SymbolFlags.Deprecated) {
+                    if (accessNode && prop.valueDeclaration?.flags & NodeFlags.Deprecated && isUncalledFunctionReference(accessNode, prop)) {
```
This is potentially `undefined & NodeFlags.Deprecated` - is that intentional?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -22043,9 +22049,8 @@ namespace ts {
-            const target = (symbol.flags & SymbolFlags.Alias ? resolveAlias(symbol) : symbol);
-            if (target.flags & SymbolFlags.Deprecated) {
-                errorOrSuggestion(/* isError */ false, node, Diagnostics._0_is_deprecated, node.escapedText as string);
+            if (declaration?.flags & NodeFlags.Deprecated && isUncalledFunctionReference(node.parent, localOrExportSymbol)) {
```
This is potentially `undefined & NodeFlags.Deprecated` - is that intentional?

sandersn commented:
@typescript-bot perf test this

sandersn commented:
@typescript-bot perf test this

sandersn commented:
Overall, there's still a good bit of variation in our perf tests. But the perf on this is still pretty good even after supporting JSX et al. I'm going to merge this since I've addressed all the comments on it and in unblocks #39388.
